server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequest.java

```diff
@@ -30,9 +34,23 @@
 
     private final BulkItemRequest[] items;
 
+    // Local, not serialized
+    private final Resources resources = new Resources();
+
     public BulkShardRequest(StreamInput in) throws IOException {
         super(in);
-        items = in.readArray(i -> i.readOptionalWriteable(inpt -> new BulkItemRequest(shardId, inpt)), BulkItemRequest[]::new);
+        boolean success = false;
+        try {
+            items = in.readArray(
+                i -> i.readOptionalWriteable(inpt -> new BulkItemRequest(shardId, inpt, resources.toRelease)),
+                BulkItemRequest[]::new
+            );
+            success = true;
+        } finally {
+            if (success == false) {
+                resources.release();
+            }
+        }
```

Comment:
Could this be simplified to:
```suggestion
        try {
            items = in.readArray(
                i -> i.readOptionalWriteable(inpt -> new BulkItemRequest(shardId, inpt, resources.toRelease)),
                BulkItemRequest[]::new
            );
        } catch(Throwable t) {
              resources.release();
        }
```
?